wlutil/wlutil.py

```diff
@@ -268,3 +278,144 @@ def copyImgFiles(img, files, direction):
                 run([sudoCmd, 'cp', '-a', os.path.normpath(mnt + f.src), f.dst])
             else:
                 raise ValueError("direction option must be either 'in' or 'out'")
+
+def getToolVersions():
+    """Detect version information for the currently enabled toolchain."""
+
+    # We run the preprocessor on a simple program to see the C-visible
+    # "LINUX_VERSION_CODE" macro
+    linuxHeaderTest = """#include <linux/version.h>
+    LINUX_VERSION_CODE
+    """
+    linuxHeaderVer = sp.run(['riscv64-unknown-linux-gnu-gcc', '-E', '-xc', '-'],
+              input=linuxHeaderTest, stdout=sp.PIPE, universal_newlines=True)
+    linuxHeaderVer = linuxHeaderVer.stdout.splitlines()[-1].strip()
+
+    # Major/minor version of the linux kernel headers included with our
+    # toolchain. This is not necessarily the same as the linux kernel used by
+    # Marshal, but is assumed to be <= to the version actually used.
+    linuxMaj = str(int(linuxHeaderVer) >> 16)
+    linuxMin = str((int(linuxHeaderVer) >> 8) & 0xFF)
+
+    # Toolchain major version
+    toolVerStr = sp.run(["riscv64-unknown-linux-gnu-gcc", "--version"],
+            universal_newlines=True, stdout=sp.PIPE).stdout
+    toolVer = toolVerStr[36]
+
+    return {'linuxMaj' : linuxMaj,
+            'linuxMin' : linuxMin,
+            'gcc' : toolVer}
+
+def checkGitStatus(submodule):
+    """Returns a dictionary representing the status of a git repo.
+
+    submodule: Path to the submodule to check. This check will be skipped if
+        the empty string is passed (the empty string is always uptodate)
+
+    Fields:
+    'sha'  : latest git commit hash (or "" if not initialized)
+    'dirty': boolean indicating whether there are uncommited changes (uninitialized repos are considered dirty)
+    'init' : boolean indiicating if the repository has been initialized
+    'rebuild' : A random number if the repo should be considered not up to date
+        for any reason (e.g. dirty==True or init==False). 0 otherwised.
+
+    This is primarily useful as an input to doit's config_changed() updtodate
+    helper which considers a workload not uptodate if some string or dictionary
+    has changed since the last time it ran. The 'sha' or 'rebuild' fields will
+    change if the repo has changed (or we can't tell if it's changed)."""
+
+    log = logging.getLogger()
+
+    if submodule == "":
```

Comment:
this is useful for having optional submodules (e.g. linux). We can just pass a default value that is trivially uptodate to represent that we don't depend on something.